### PR TITLE
Newapp crd 311

### PR DESCRIPTION
**Pull request by gabemontero**

### CODE REVIEW

**Comment by gabemontero** [comment]
interesting, though not necessarily surprising ... the error message with our new test case is different on 3.11 than it was on 4.0:

```
=== BEGIN TEST CASE ===
test/cmd/newapp.sh:109: executing 'oc new-app -f test/testdata/newapp/template-with-crd.yaml' expecting failure and text 'error: The template contained an object type unknown to '
FAILURE after 0.305s: test/cmd/newapp.sh:109: executing 'oc new-app -f test/testdata/newapp/template-with-crd.yaml' expecting failure and text 'error: The template contained an object type unknown to ': the output content test failed
Standard output from the command:
--> Deploying template "cmd-newapp/template-with-crd" for "test/testdata/newapp/template-with-crd.yaml" to project cmd-newapp

--> Creating resources ...
--> Failed

Standard error from the command:
    error: CustomResourceDefinition.apiextensions.k8s.io "template-with-crd" is invalid: [metadata.name: Invalid value: "template-with-crd": must be spec.names.plural+"."+spec.group, spec.group: Required value, spec.versions: Invalid value: []apiextensions.CustomResourceDefinitionVersion(nil): must have exactly one version marked as storage version, spec.names.plural: Required value, spec.names.singular: Required value, spec.names.kind: Required value, spec.names.listKind: Required value, status.storedVersions: Invalid value: []string(nil): must have at least one stored version]
=== END TEST CASE ===
```

will need to update

**Comment by gabemontero** [comment]
OK figured out the 3.11.x vs. 4.x diff.

It was Juan's https://github.com/openshift/origin/commit/0cc42c851a15fd4e4fb875a4658b097b7da7c6ad#diff-3110d88fd0c340544c2b2a479b7fc78d via PR https://github.com/openshift/origin/pull/20645 for externalizing the newapp command that injected the attempt at decoding and uncovered the inability for new-app to understand CRDs much earlier and in a more explicit way than the 3.11 code base currently does.

That's a huge commit for starters and might have lower level dependencies on top of that (I haven't sifted through all that - I'm hoping @soltysh can chime in here).

Assuming my instincts are correct there, I'm going to see if I can just lift https://github.com/openshift/origin/blob/master/pkg/oc/lib/newapp/cmd/newapp.go#L544-L548 ... I'll copy / paste those lines here:

```
		resultObjects, err := templateObjectsToAppObjects(result.Objects)
		if err != nil {
			return "", nil, err
		}

```

Where `templateObjectsToAppObjects` attempts to decode:

```
		resultObjects, err := templateObjectsToAppObjects(result.Objects)
		if err != nil {
			return "", nil, err
		}

```

If the `runtime` and `scheme.Codecs` dependencies prove untenable, then we are talking intercepting the error that occurs much further down the processing line and sorting out how to translate the error in https://github.com/openshift/origin/pull/22654#issuecomment-486269882 as being "we do not understand CRDs"

**Comment by gabemontero** [comment]
had lunch with @deads2k and he said attempting just the decode in 3.11 was untenable....can't expect reliable results.

**Comment by gabemontero** [comment]
@adambkaplan fyi ^^ https://github.com/openshift/origin/pull/22654#issuecomment-486737092 and https://github.com/openshift/origin/pull/22654#issuecomment-486722914

**Comment by gabemontero** [comment]
ok @soltysh @adambkaplan I have a hack up that gets what we agreed to message wise in 3.11

the hack is at least small ... I'll refrain from attempting to provide any other positive characterizations :-)

ptal

File: pkg/oc/cli/newapp/newapp.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ import (
 	"github.com/openshift/origin/pkg/cmd/util/print"
 	imageapi "github.com/openshift/origin/pkg/image/apis/image"
 	imageclientinternal "github.com/openshift/origin/pkg/image/generated/internalclientset"
-	"github.com/openshift/origin/pkg/oc/lib/newapp"
+	generate "github.com/openshift/origin/pkg/oc/lib/newapp"
```
Comment (adambkaplan) [inline pkg/oc/cli/newapp/newapp.go]: Where is this coming from? Feels like it would break compilation

**Comment by gabemontero** [comment]
On Thu, Apr 25, 2019 at 9:34 PM Adam Kaplan <notifications@github.com>
wrote:

> *@adambkaplan* requested changes on this pull request.
> ------------------------------
>
> In pkg/oc/cli/newapp/newapp.go
> <https://github.com/openshift/origin/pull/22654#discussion_r278784624>:
>
> > @@ -47,7 +47,7 @@ import (
>  	"github.com/openshift/origin/pkg/cmd/util/print"
>  	imageapi "github.com/openshift/origin/pkg/image/apis/image"
>  	imageclientinternal "github.com/openshift/origin/pkg/image/generated/internalclientset"
> -	"github.com/openshift/origin/pkg/oc/lib/newapp"
> +	generate "github.com/openshift/origin/pkg/oc/lib/newapp"
>
> Where is this coming from? Feels like it would break compilation
>
vscode golang tooling picked it up;  there is some non-standard package
names in the go files in those dirs;  turns out this clears up some
ambiguity with the surrounding imports so I left it.  It definitely
compiles.

**Comment by gabemontero** [comment]
bump @soltysh @adambkaplan

File: pkg/oc/cli/newapp/newapp.go (hunk 1 above)
Comment (bparees) [reply pkg/oc/cli/newapp/newapp.go]: ah...  the files in oc/lib/newapp are actually in the "generate" package.  who the heck.......

Comment (gabemontero) [reply pkg/oc/cli/newapp/newapp.go]: yep that's right re: https://github.com/openshift/origin/pull/22654#issuecomment-486912132

**Pull request merged**